src/System.Binary/System/Binary/UnsafeUtilities.cs

```diff
@@ -64,5 +64,55 @@ internal static class UnsafeUtilities
                 return Unsafe.Read<T>(val);
             }
         }
+
+        public static short ReverseEndianness(short value)
+        {
+            return (short)((value & 0x00FF) << 8 | (value & 0xFF00) >> 8);
+        }
+
+        public static int ReverseEndianness(int value)
+        {
+            return (value & 0x000000FF) << 24 |
+                (value & 0x0000FF00) << 8 |
+                (value & 0x00FF0000) >> 8 |
+                (int)(((uint)value & 0xFF000000) >> 24);
+        }
+
+        public static long ReverseEndianness(long value)
+        {
+            return (value & 0x00000000000000FFL) << 56 |
```

Comment:
Is there a bit perf hit if we share the implementation between ulong and long? (inlining the shared part)

Reply:
I will check.

Reply:
There is a ~10% performance difference (if I change it to something like what's shown below):
![image](https://user-images.githubusercontent.com/6527137/30677231-6e633cf8-9e3f-11e7-93b5-20af9abb79e2.png)

```C#
       public static ulong ReverseEndianness(ulong value)
        {
            return SharedHelper(value) | (value & 0xFF00000000000000UL) >> 56;
        }

        public static long ReverseEndianness(long value)
        {
            return (long)SharedHelper((ulong)value) | (long)(((ulong)value & 0xFF00000000000000L) >> 56);
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        private static ulong SharedHelper(ulong value)
        {
            return (value & 0x00000000000000FFUL) << 56 |
                (value & 0x000000000000FF00UL) << 40 |
                (value & 0x0000000000FF0000UL) << 24 |
                (value & 0x00000000FF000000UL) << 8 |
                (value & 0x000000FF00000000UL) >> 8 |
                (value & 0x0000FF0000000000UL) >> 24 |
                (value & 0x00FF000000000000UL) >> 40;
        }
```

Reply:
Thinking about it some more and looking at the disassembly, theoretically there shouldn't be any performance difference.

The only difference I see is the use of shr vs sar instructions (signed divide vs unsigned divide) for the `ReverseEndianness(long value)`.

![image](https://user-images.githubusercontent.com/6527137/30712045-6ea8471e-9ebf-11e7-9ad9-d88a3538cf4f.png)

I don't know why the results consistently differ though:
![image](https://user-images.githubusercontent.com/6527137/30712837-36a3b6ac-9ec2-11e7-8d6d-6f7b4133cab5.png)

@jkotas, any ideas why we would see a performance difference here?

@jorive, here is what my benchmark looks like. Is there something I can do to improve how I write the benchmark?
```C#
private const int TestArrayLength = 1000000;
private static byte[] myArray = new byte[TestArrayLength * 8];
volatile static int tempVal;

[Benchmark]
private static void MyTestAOld()
{
    Span<byte> source = GetArray();
            
    foreach (var iteration in Benchmark.Iterations)
    {
        tempVal = 0;
        using (iteration.StartMeasurement())
        {
            for (int j = 0; j < TestArrayLength; j++)
            {
                tempVal ^= (int)source.Slice(j * 8).ReadInt64BigEndianOld();
            }
        }
        Assert.Equal(tempVal, 1173436863);
    }
}

[Benchmark]
private static void MyTestBNew()
{
    Span<byte> source = GetArray();
            
    foreach (var iteration in Benchmark.Iterations)
    {
        tempVal = 0;
        using (iteration.StartMeasurement())
        {
            for (int j = 0; j < TestArrayLength; j++)
            {
                tempVal ^= (int)source.Slice(j * 8).ReadInt64BigEndian();
            }
        }
        Assert.Equal(tempVal, 1173436863);
    }
}

private static Span<byte> GetArray()
{
    Span<byte> source = myArray;
    Random random = new Random(42);
    for (int i = 0; i < TestArrayLength; i++)
    {
        int temp = random.Next();
        source.Slice(i * 8).WriteInt64BigEndian(temp);
    }
    return source;
}
```

Reply:
> private const int TestArrayLength = 1000000;

Iterating over 4MB array is processor cache bound. 4MBs will definitely not fit into L1. It may fit into L2 or L3 depending on what kind of machine you are doing the measurements on.

My guess is that the exact array location in the memory may be influencing the benchmark results because of iteractions with the cache. You may want to try to run both benchmarks on the exact same array.

Reply:
@erozenfeld It may be nice to pattern match this as `bswap`.

Reply:
>  You may want to try to run both benchmarks on the exact same array.

I am using the same array. I wrap a span around `private static byte[] myArray = new byte[TestArrayLength * 8];` that I pass to the methods being measured.

I can try a smaller array (with a high inner iteration count) that would fit into L1 would (4-16 KB be fine)?

Reply:
@jkotas I opened https://github.com/dotnet/coreclr/issues/14122 for bswap support in the jit.